airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/TypeAndDedupeOperationValve.java

```diff
@@ -0,0 +1,153 @@
+      FIVE_MINUTES_MILLIS,
+      TEN_MINUTES_MILLIS,
+      FIFTEEN_MINUTES_MILLIS);
+
+  // Constantly getting the system time adds a bit of overhead, adding a minimum record count
+  // To reduce calls for system time
+  private static final int MINIMUM_RECORD_INTERVAL = 100;
+
+  private static final Supplier<Long> SYSTEM_NOW = () -> System.currentTimeMillis();
+
+  private ConcurrentHashMap<AirbyteStreamNameNamespacePair, Integer> incrementalIndex;
+
+  private final Supplier<Long> nowness;
+  private ConcurrentHashMap<AirbyteStreamNameNamespacePair, Long> recordCounts;
+
+  public TypeAndDedupeOperationValve() {
+    this(SYSTEM_NOW);
+  }
+
+  /**
+   * This constructor is here because mocking System.currentTimeMillis() is a pain :(
+   *
+   * @param nownessSupplier Supplier which will return a long value representing now
+   */
+  public TypeAndDedupeOperationValve(Supplier<Long> nownessSupplier) {
+    super();
+    incrementalIndex = new ConcurrentHashMap<>();
+    recordCounts = new ConcurrentHashMap<>();
+    this.nowness = nownessSupplier;
+  }
+
+  @Override
+  public Long put(final AirbyteStreamNameNamespacePair key, final Long value) {
+    if (!incrementalIndex.containsKey(key)) {
+      incrementalIndex.put(key, 0);
+    }
+    if (!recordCounts.containsKey(key)) {
+      recordCounts.put(key, 1l);
+    }
+    return super.put(key, value);
+
+  }
+
+  /**
+   * Adds a stream specific timestamp to track type and dedupe operations
+   *
+   * @param key the AirbyteStreamNameNamespacePair to track
+   */
+  public void addStream(final AirbyteStreamNameNamespacePair key) {
+    put(key, nowness.get());
+  }
+
+  /**
+   * Whether we should type and dedupe at this point in time for this particular stream.
+   *
+   * @param key the stream in question
+   * @return a boolean indicating whether we have crossed the interval threshold for typing and
+   *         deduping.
+   */
+  public boolean readyToTypeAndDedupe(final AirbyteStreamNameNamespacePair key) {
+    if (!containsKey(key)) {
+      return false;
+    }
+    recordCounts.put(key, recordCounts.get(key) + 1);
+    if (recordCounts.get(key) % MINIMUM_RECORD_INTERVAL == 0) {
+      return nowness.get() - get(key) > typeAndDedupeIncreasingIntervals.get(incrementalIndex.get(key));
+    }
+    return false;
+  }
+
+  /**
+   * Increment the interval at which typing and deduping should occur for the stream, max out at last
+   * index of {@link TypeAndDedupeOperationValve#typeAndDedupeIncreasingIntervals}
+   *
+   * @param key the stream to increment the interval of
+   * @return the index of the typing and deduping interval associated with this stream
+   */
+  public int incrementInterval(final AirbyteStreamNameNamespacePair key) {
+    if (incrementalIndex.get(key) < typeAndDedupeIncreasingIntervals.size() - 1) {
+      incrementalIndex.put(key, incrementalIndex.get(key) + 1);
+    }
+    return incrementalIndex.get(key);
+  }
+
+  /**
+   * Meant to be called after
```

Comment:
why can't this happen inside readyToTypeAndDedupe?

Reply:
I liked the separation of concerns here. Readiness to type and dedupe and increasing the interval aren't necessarily coupled even if I've stated they're meant to be 😄.